### PR TITLE
feat(gas-oracle): support gas token volatile exchange rate

### PR DESCRIPTION
**Pull request by yiweichi**
### Purpose or design rationale of this PR

*Describe your change. Make sure to answer these three questions: What does this PR do? Why does it do it? How does it do it?*

Since we support using alternative gas on L2,  we should include l2_gas_token/l1_gas_token price information in to gas price both ways. So here we support retrieve l2_gas_token/l1_gas_token price from binance_api, and includeit in the fee data if alternative gas token enabled.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

File: rollup/internal/config/relayer.go [hunk 1]
```diff
@@ -68,12 +95,22 @@ type RelayerConfig struct {
 	FinalizeBundleWithoutProofTimeoutSec uint64 `json:"finalize_bundle_without_proof_timeout_sec"`
 }
 
+// AlternativeGasTokenConfig The configuration for handling token exchange rates when updating the gas price oracle.
```
Comment (NazariiDenha) [inline rollup/internal/config/relayer.go:71]: lgtm

The one thing is maybe add here description which token symbol pair should be chosen. For example, if L1 gas token is `ETH` and L2 gas token `SCR`, which ticker should be used `ETHSCR` or `SCRETH`. This is not obvious from the first look

**Review by NazariiDenha — DISMISSED** [review]

File: rollup/internal/config/relayer.go (hunk 1 above)
Comment (yiweichi) [reply rollup/internal/config/relayer.go:71]: thanks, added notation for this.

File: rollup/internal/controller/relayer/l1_relayer.go [hunk 2]
```diff
@@ -151,6 +153,27 @@ func (r *Layer1Relayer) ProcessGasPriceOracle() {
...
+				exchangeRate, err = rutils.GetExchangeRateFromBinanceApi(r.cfg.GasOracleConfig.AlternativeGasTokenConfig.TokenSymbolPair)
+				if err != nil {
+					log.Error("Failed to get gas token exchange rate from Binance api", "tokenSymbolPair", r.cfg.GasOracleConfig.AlternativeGasTokenConfig.TokenSymbolPair, "err", err)
+					return
+				}
+			default:
+				log.Error("Invalid alternative gas token mode", "mode", r.cfg.GasOracleConfig.AlternativeGasTokenConfig.Mode)
+				return
+			}
+			baseFee = uint64(math.Ceil(float64(baseFee) / exchangeRate))
```
Comment (colinlyguo) [inline rollup/internal/controller/relayer/l1_relayer.go:176]: could `exchangeRate` be 0 in some cases (e.g. bad query result)? may add some validity checks in `exchangeRate`, `blobBaseFee`, `baseFee`. e.g. > 0 and < 2^256.

**Review by colinlyguo — DISMISSED** [review]
lgtm.

File: rollup/internal/config/relayer.go [hunk 3]
```diff
@@ -68,12 +95,22 @@ type RelayerConfig struct {
...
+type AlternativeGasTokenConfig struct {
+	Enabled           bool             `json:"enabled"`
+	Mode              ExchangeRateMode `json:"mode"`
+	FixedExchangeRate float64          `json:"fixed_exchange_rate"`
```
Comment (colinlyguo) [inline rollup/internal/config/relayer.go]:
```suggestion
	FixedExchangeRate float64          `json:"fixed_exchange_rate"` // fixed exchange rate of L2 gas token / L1 gas token
```

File: rollup/internal/controller/relayer/l1_relayer.go (hunk 2 above)
Comment (yiweichi) [reply rollup/internal/controller/relayer/l1_relayer.go:176]: thanks, added validity check for exchangeRate.

**Review by colinlyguo — DISMISSED** [review]

File: rollup/internal/controller/relayer/l1_relayer.go [hunk 4]
```diff
@@ -17,6 +17,8 @@ import (
 	"scroll-tech/common/types"
 	"scroll-tech/common/utils"
 
+	rutils "scroll-tech/rollup/internal/utils"
+
```
Comment (georgehao) [inline rollup/internal/controller/relayer/l1_relayer.go]: delete empty line

File: rollup/internal/config/relayer.go [hunk 5]
```diff
@@ -68,12 +95,22 @@ type RelayerConfig struct {
 	FinalizeBundleWithoutProofTimeoutSec uint64 `json:"finalize_bundle_without_proof_timeout_sec"`
 }
 
+// AlternativeGasTokenConfig The configuration for handling token exchange rates when updating the gas price oracle.
+type AlternativeGasTokenConfig struct {
+	Enabled           bool             `json:"enabled"`
+	Mode              ExchangeRateMode `json:"mode"`
```
Comment (georgehao) [inline rollup/internal/config/relayer.go]: string is enough

File: rollup/internal/controller/relayer/l1_relayer.go [hunk 6]
```diff
@@ -151,6 +153,31 @@ func (r *Layer1Relayer) ProcessGasPriceOracle() {
...
+				}
+			default:
+				log.Error("Invalid alternative gas token mode", "mode", r.cfg.GasOracleConfig.AlternativeGasTokenConfig.Mode)
+				return
+			}
+			if exchangeRate == 0 {
+				log.Error("Invalid exchange rate", "exchangeRate", exchangeRate)
+				return
+			}
+			baseFee = uint64(math.Ceil(float64(baseFee) / exchangeRate))
```
Comment (georgehao) [inline rollup/internal/controller/relayer/l1_relayer.go:176]: What's rationale of `baseFee/exchangeRate`

File: rollup/internal/utils/exchange_rate.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/scroll-tech/go-ethereum/log"
+)
+
+var BinanceApiEndpoint string = "https://api.binance.com/api/v3/ticker/price?symbol=%s"
+
+type BinanceResponse struct {
+	Price string `json:"price"`
+}
+
+func GetExchangeRateFromBinanceApi(tokenSymbolPair string) (float64, error) {
```
Comment (georgehao) [inline rollup/internal/utils/exchange_rate.go]: maybe add some retry logic of get BinanceApiEndpoint

File: rollup/internal/controller/relayer/l1_relayer.go (hunk 6 above)
Comment (yiweichi) [reply rollup/internal/controller/relayer/l1_relayer.go:176]: exchangeRate is the number of native token on L1 required to exchange for 1 native token on L2.
so when updating baseFee to layer2, we need to convert baseFee unit from L1 gas token to L2 gas token.

File: rollup/internal/config/relayer.go (hunk 5 above)
Comment (yiweichi) [reply rollup/internal/config/relayer.go]: thanks, changed to string.

File: rollup/internal/utils/exchange_rate.go (hunk 7 above)
Comment (yiweichi) [reply rollup/internal/utils/exchange_rate.go]: thanks, added retry logic.

File: rollup/internal/utils/exchange_rate.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/log"
+)
+
+var BinanceApiEndpoint string = "https://api.binance.com/api/v3/ticker/price?symbol=%s"
```
Comment (colinlyguo) [inline rollup/internal/utils/exchange_rate.go:14]: Could this URL template changed later? if so, can move it to config, if not frequent, can keep it here.

**Review by colinlyguo — DISMISSED** [review]

File: rollup/internal/utils/exchange_rate.go (hunk 8 above)
Comment (yiweichi) [reply rollup/internal/utils/exchange_rate.go:14]: I think not frequent. btw, I've tried to put it to config, the ci failed, it says: http Get a arbitrary url is not allowed.

**Review by georgehao — APPROVED** [review]

**Review by colinlyguo — APPROVED** [review]

**Pull request merged**